File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
# 대각선 방향으로 한 칸 오른쪽 or 왼쪽
# 거쳐간 숫자의 최댓값을 return -> dp 문제 
def solution(triangle):
    for i in range(1, len(triangle)):
        triangle[i][0] += triangle[i-1][0]
        triangle[i][-1] += triangle[i-1][-1]
        
    for i in range(2, len(triangle)):
        for j in range(1, i):
            triangle[i][j] += max(triangle[i-1][j-1], triangle[i-1][j])
    
    
    
    return max(triangle[-1])

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
import java.util.*;
class Solution {
    public int solution(int[][] triangle) {
        int[][] memo = new int[500][500];
        memo[0][0] = triangle[0][0];
        
        for(int i = 1; i < triangle.length; i++) {
            for(int j = 0; j <= i; j++) {
                if(j == 0) 
                    memo[i][j] = memo[i-1][0] + triangle[i][j];
                else if(j == i)
                    memo[i][j] = memo[i-1][j-1] + triangle[i][j];
                else 
                    memo[i][j] = Math.max(memo[i-1][j-1], memo[i-1][j]) + triangle[i][j];
            }
        }
        
        int answer = 0;
        for(int n : memo[triangle.length - 1])
            answer = Math.max(n, answer);
        
        return answer;
    }
}